src/test/connect-test.c

```diff
@@ -0,0 +1,444 @@
+        if (r < 0) {
+                fprintf(stderr, "IPv4 client bind failed: %s\n", strerror(errno));
+                goto end;
+        }
+
+        addr_4.sin_port = htons(port);
+
+        /* try to connect both sockets */
+        r = connect(fd_4, (struct sockaddr *) &addr_4, sizeof(addr_4));
+        if (r == 0)
+                ctx.ipv4 = CONNECTION_SUCCEEDED;
+        else if (r < 0 && errno == EINPROGRESS) {
+                r = sd_event_add_io(e, &event_source, fd_4, EPOLLOUT, conn_handler_4, &ctx);
+                if (r < 0)
+                        goto end;
+        } else {
+                fprintf(stderr, "IPv4 connection failed: %s\n", strerror(r));
+                goto end;
+        }
+
+        fd_6 = socket(AF_INET6, SOCK_STREAM | SOCK_NONBLOCK, 0);
+        if (fd_6 < 0) {
+                r = -errno;
+                goto end;
+        }
+
+        r = setsockopt(fd_6, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(reuseaddr));
+        if (r < 0)
+                goto end;
+
+        addr_6.sin6_family = AF_INET6;
+        addr_6.sin6_port = htons(port+1);
+        r = inet_pton(AF_INET6, "::1", &addr_6.sin6_addr);
+        if (r < 0)
+                goto end;
+
+        r = bind(fd_6, (struct sockaddr *) &addr_6, sizeof(addr_6));
+        if (r < 0) {
+                fprintf(stderr, "IPv6 client bind failed: %s\n", strerror(r));
+                goto end;
+        }
+
+        addr_6.sin6_port = htons(port);
+
+        r = connect(fd_6, (struct sockaddr *) &addr_6, sizeof(addr_6));
+        if (r == 0)
+                ctx.ipv6 = CONNECTION_SUCCEEDED;
+        else if (r < 0 && errno == EINPROGRESS) {
+                r = sd_event_add_io(e, &event_source, fd_6, EPOLLOUT, conn_handler_6, &ctx);
+                if (r < 0)
+                        goto end;
+        } else {
+                fprintf(stderr, "IPv6 connection failed: %s\n", strerror(errno));
+                goto end;
+        }
+
+        if (ctx.ipv4 != CONNECTION_UNDEFINED && ctx.ipv6 != CONNECTION_UNDEFINED) {
+                goto end;
+        }
+
+        /* add the timeout */
+        usec = now(CLOCK_MONOTONIC) + (timeout * USEC_PER_MSEC);
+        if (sd_event_add_time(e, &event_source, CLOCK_MONOTONIC, usec, 0, timeout_handler, &ctx))
+                return -1;
+
+        /*
+         * Run the event loop -- wait for connections to complete or timeout to
+         * trigger.
+         */
+        r = sd_event_loop(e);
+        if (r < 0) {
+                fprintf(stderr, "Event loop error: %s\n", strerror(-r));
+                goto end;
+        }
+
+end:
+        if (event_source)
+                sd_event_source_unref(event_source);
+
+        if (e)
+                sd_event_unref(e);
+
+        if (fd_4 >= 0)
+                close(fd_4);
+
+        if (fd_6 >= 0)
+                close(fd_6);
+
+        if (listen_fd_4 >= 0)
+                close(listen_fd_4);
+
+        if (listen_fd_6 >= 0)
+                close(listen_fd_6);
```

Comment:
consider using cleanup macors for all of the above. i.e. declare the fds with _cleanup_close_ and so on, so that this can go